internal/compiler/object_tree.rs

```diff
@@ -912,13 +922,73 @@ impl Element {
                 name,
                 PropertyDeclaration {
                     property_type: Type::Callback { return_type, args },
-                    node: Some(Either::Right(sig_decl)),
+                    node: Some(sig_decl.into()),
                     visibility: PropertyVisibility::InOut,
                     ..Default::default()
                 },
             );
         }
 
+        for func in node.Function() {
+            let name =
+                unwrap_or_continue!(parser::identifier_text(&func.DeclaredIdentifier()); diag);
+
+            let PropertyLookupResult {
+                resolved_name: existing_name,
+                property_type: maybe_existing_prop_type,
+                ..
+            } = r.lookup_property(&name);
+            if !matches!(maybe_existing_prop_type, Type::Invalid) {
```

Comment:
yes, currently private properties cannot be overwritten.  I was thinking of allowing it but maybe later as the compiler will need to be changed a lot

Reply:
Shall I create a ticket for this? IMO we should fix this before 1.0 (and I volunteer to do it, unless you'd like to).

Reply:
You can create a ticket, yes.  But IMHO this doesn't need to be done for 1.0.  Because adding this will not break any code and will only let code that compile.

I think it is quite some work to do, because of inlining and other passes that assumes that everything override the base.

Reply:
I think it should be fixed before 1.0 because adding new private properties breaks existing code. Imagine the above example where "foo" doesn't exist in `Base`. Later the implementation is re-worked and the designer decides that they need `foo`. They can add it, declare it private and it's all good. Only later, when the `.slint` file is compiled in another project where `Derived` exists, then the build breaks.

I mean yeah, one could argue that we're not making anything worse than it is today, but at the same time I feel this is part of the promise of "private", part of what would make it useful.

Reply:
adding new private or public properties may breaks existing code, yes.

But this is why we shouldn't promote inheritence

Reply:
For public I think that's "obvious", but for private it isn't - at least I think it's most certainly not what users with a programming background will expect.

> I think it is quite some work to do, because of inlining and other passes that assumes that everything override the base.

Right, inlining is what made wonder about this in the context of functions, and then I realised it's global :). That said, we already have a similar implementation in `move_declarations` after renaming properties and fixing up references, don't we?

Reply:
I think we agree that this needs to be changed.
Even for public field, we could allow overriding them (maybe with a warning).

Anyway, feel free to tackle the issue, but this will require big refactoring in the compiler (which may be a good thing anyway).

I just don't see the urgency to do it, because this is not a frequent problem, and we can solve the problem later.

> For public I think that's "obvious", but for private it isn't - at least I think it's most certainly not what users with a programming background will expect.

In C++ it is a breaking change to add private members  Some examples: https://godbolt.org/z/rn3r18Tqb

> Right, inlining is what made wonder about this in the context of functions, and then I realised it's global :). That said, we already have a similar implementation in move_declarations after renaming properties and fixing up references, don't we?

Yeah, we could rename all private properties  to something else, that would be a quick hack to fix the problem.
But then we still have to have some code to diagnose usage of private property so we can tell the user the property exist but is private if one tries to use a private property.

Reply:
> I think we agree that this needs to be changed. Even for public field, we could allow overriding them (maybe with a warning).
> 
> Anyway, feel free to tackle the issue, but this will require big refactoring in the compiler (which may be a good thing anyway).

Ok. I'll create a ticket, assign it to myself.

> I just don't see the urgency to do it, because this is not a frequent problem, and we can solve the problem later.

It's true, this can be fixed later. I just feel that it makes the feature "less complete".

> > For public I think that's "obvious", but for private it isn't - at least I think it's most certainly not what users with a programming background will expect.
> 
> In C++ it is a breaking change to add private members Some examples: https://godbolt.org/z/rn3r18Tqb

Yes, and C++ is a great example of what happens when the encapsulation started out leaky. I think a new C++ should behave as if "private" fields were in a d-pointer.

JavaScript gets it right:

 https://www.typescriptlang.org/play?#code/MYGwhgzhAEAKCuAjEBLYAhSBTaBvAUNEdAGYD2ZAXNBAC4BOKAdgObQC80ARItl4cQAOjJrQAUASjwDixYGSYQyILADoQZFmNoALFBFXkyEgNwzoAX3xX8oSDATI0AESyMAblgAm0LAA9aLCYvByRUDGxpWSNqOhE2Ti4vNxRPL34bFVpoQSQOaCYsAHc4MJcUtLNcxFVhZnFTfFtwKDgPMEDMCBwCWQBiGJoGZgTuOvcOrH5ZOtFJKNk5BSUVdU1tPQMBikbZKxs7Vth2wNcPb18AoJC21MmunvNBuJH8pIrvDKasnI98wpKxzupw+XiqHlqIgaJiAA

(TypeScript transpiles to the use of a WeakMap per instance, but native private classes behave the same)

> > Right, inlining is what made wonder about this in the context of functions, and then I realised it's global :). That said, we already have a similar implementation in move_declarations after renaming properties and fixing up references, don't we?
> 
> Yeah, we could rename all private properties to something else, that would be a quick hack to fix the problem. But then we still have to have some code to diagnose usage of private property so we can tell the user the property exist but is private if one tries to use a private property.

Right, that may not be the right path. Wasn't quite thinking of a quick hack anyway, but your argument is probably an argument in favor of storing these fields in a separate data structure.